### PR TITLE
Online patient lookup API

### PR DESCRIPTION
**Pull request by ssrihari**
**Story card:** [ch3257](https://app.clubhouse.io/simpledotorg/story/3257/add-a-new-api-for-looking-up-patients-by-their-identifier)

## Because
This implements the server side of the online patient lookup feature.

- [ADR](https://github.com/simpledotorg/simple-server/blob/master/doc/arch/015-online-patient-lookup.md)
- [PRD](https://docs.google.com/document/d/1q6cppByQULfh3_mMXC4BJpiNN9Uc_awA6rreeEtUBaM/edit#)

## This addresses

- API implementation, including schema, transformation
- Access restriction to state level data
- Audit logging
- Rate limiting

## Test instructions

- Make requests to `/api/v4/patients/{identifier}`
- Ensure it works for a single patient, and then multiple patients with the same identifier
- Test rate limiting end to end
- Test that audit logs are written to file in a separate file

### CODE REVIEW

File: app/controllers/api/v4/patients_controller.rb [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        user_id: current_user.id,
+        facility_id: current_facility.id,
+        identifier: params[:identifier],
+        patient_ids: patients.pluck(:id),
+        time: Time.current
+      }.to_json
+    )
+  end
+
+  def validate_identifier_presence
```
Comment (prabhanshuguptagit) [inline app/controllers/api/v4/patients_controller.rb:66]: We could do 
```
def lookup_params
  params.require(:identifier)
end
```
This also returns a bad request for missing params, I think.

File: app/controllers/api/v4/patients_controller.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        blood_pressures: patient.blood_pressures.map { |blood_pressure| Api::V3::BloodPressureTransformer.to_response(blood_pressure) },
+        blood_sugars: patient.blood_sugars.map { |blood_sugar| Api::V4::BloodSugarTransformer.to_response(blood_sugar) },
+        prescription_drugs: patient.prescription_drugs.map { |prescription_drug| Api::V3::PrescriptionDrugTransformer.to_response(prescription_drug) },
+        retention: retention(patient)
+      }
+    )
+  end
+
+  def retention(patient)
+    if current_sync_region.syncable_patients.exists?(patient.id)
```
[the remark this answers is not in the document]
Comment (prabhanshuguptagit) [reply app/controllers/api/v4/patients_controller.rb:42]: Also, what happens if the patient gets synced from the regular sync API after being synced from this endpoint?

**Review by prabhanshuguptagit — COMMENTED** [review]
Looking slick ✨

File: app/controllers/api/v4/patients_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ssrihari) [reply app/controllers/api/v4/patients_controller.rb:42]: I'll add this to the ADR, that we need to handle overwriting of temporary -> permanent when a temporary-patient is synced via the regular sync API.

File: app/controllers/api/v4/patients_controller.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        patients: patients.map { |patient| transform_to_response(patient) }
+      }, mode: :compat),
+      status: :ok
+    )
+  end
+
+  private
+
+  def transform_to_response(patient)
+    Api::V3::PatientTransformer.to_nested_response(patient).merge(
```
[the remark this answers is not in the document]
Comment (ssrihari) [reply app/controllers/api/v4/patients_controller.rb:29]: Thanks for this suggestion, I'll do this.

File: app/controllers/api/v4/patients_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ssrihari) [reply app/controllers/api/v4/patients_controller.rb:42]: If ^ this difficult or complex to do, we can consider making all retentions temporary.

File: app/controllers/api/v4/patients_controller.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (ssrihari) [reply app/controllers/api/v4/patients_controller.rb]: Done in f0c4558f999d61baa4a4c130553663e9c855a5a3.

File: app/controllers/api/v4/patients_controller.rb (hunk 1 above)
Comment (ssrihari) [reply app/controllers/api/v4/patients_controller.rb]: Addressed in f0c4558f999d61baa4a4c130553663e9c855a5a3.

File: app/controllers/api/v4/patients_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ssrihari) [reply app/controllers/api/v4/patients_controller.rb:28]: Added a note to the ADR. From our conversation, leaning into keeping the `RETENTION_TYPE` for now.

**Review by ssrihari — COMMENTED** [review]
Addressed review comments @prabhanshuguptagit.

**Review by prabhanshuguptagit — APPROVED** [review]
LGTM 🚀

**Pull request merged**